File: bindings/csharp/tests/SmokeTest.cs
// ============================================================================
// UltrafastSecp256k1 -- C# Binding Smoke Test (Golden Vectors)
// ============================================================================
// Verifies FFI boundary correctness using deterministic known-answer tests.
// Runs in <2 seconds.
//
// Usage:
//   dotnet run --project bindings/csharp/SmokeTest
// ============================================================================

using System;
using System.Linq;

namespace UltrafastSecp256k1.SmokeTest
{
    class Program
    {
        static int passed = 0, failed = 0;

        // ── Golden Vectors ──────────────────────────────────────────────

        static readonly byte[] KNOWN_PRIVKEY = HexToBytes(
            "0000000000000000000000000000000000000000000000000000000000000001");

        static readonly byte[] KNOWN_PUBKEY = HexToBytes(
            "0279BE667EF9DCBBAC55A06295CE870B07029BFCDB2DCE28D959F2815B16F81798");

        static readonly byte[] KNOWN_XONLY = HexToBytes(
            "79BE667EF9DCBBAC55A06295CE870B07029BFCDB2DCE28D959F2815B16F81798");

        static readonly byte[] SHA256_EMPTY = HexToBytes(
            "E3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B934CA495991B7852B855");

        static readonly byte[] MSG32 = new byte[32];
        static readonly byte[] AUX32 = new byte[32];

        // ── Tests ───────────────────────────────────────────────────────

        static void Test(string name, Action fn)
        {
            try
            {
                fn();
                Console.WriteLine($"  [OK] {name}");
                passed++;
            }
            catch (Exception e)
            {
                Console.WriteLine($"  [FAIL] {name}: {e.Message}");
                failed++;
            }
        }

        static void Main()
        {
            Console.WriteLine("UltrafastSecp256k1 C# Smoke Test");
            Console.WriteLine(new string('=', 60));

            Test("ctx_create_abi", () =>
            {
                using var ctx = new Ufsecp.Ufsecp();
                var abi = ctx.AbiVersion;
                Assert(abi >= 1, $"ABI {abi} < 1");
            });

            Test("pubkey_create_golden", () =>
            {
                using var ctx = new Ufsecp.Ufsecp();
                var pub = ctx.PubkeyCreate(KNOWN_PRIVKEY);
                AssertEqual(pub, KNOWN_PUBKEY, "compressed pubkey");
            });

            Test("pubkey_xonly_golden", () =>
            {
                using var ctx = new Ufsecp.Ufsecp();
                var xonly = ctx.PubkeyXonly(KNOWN_PRIVKEY);
                AssertEqual(xonly, KNOWN_XONLY, "x-only pubkey");
            });

            Test("ecdsa_sign_verify", () =>
            {
                using var ctx = new Ufsecp.Ufsecp();
                var sig = ctx.EcdsaSign(MSG32, KNOWN_PRIVKEY);
                Assert(sig.Length == 64, "sig length");
                ctx.EcdsaVerify(MSG32, sig, KNOWN_PUBKEY);

                // Mutated → fail
                var bad = (byte[])sig.Clone();
                bad[0] ^= 0x01;
                bool threw = false;
                try { ctx.EcdsaVerify(MSG32, bad, KNOWN_PUBKEY); }
                catch { threw = true; }
                Assert(threw, "mutated sig should fail");
            });

            Test("schnorr_sign_verify", () =>
            {
                using var ctx = new Ufsecp.Ufsecp();
                var sig = ctx.SchnorrSign(MSG32, KNOWN_PRIVKEY, AUX32);
                Assert(sig.Length == 64, "sig length");
                ctx.SchnorrVerify(MSG32, sig, KNOWN_XONLY);
            });

            Test("ecdsa_recover", () =>
            {
                using var ctx = new Ufsecp.Ufsecp();
                var (sig, recid) = ctx.EcdsaSignRecoverable(MSG32, KNOWN_PRIVKEY);
                Assert(recid >= 0 && recid <= 3, "recid range");
                var pub = ctx.EcdsaRecover(MSG32, sig, recid);
                AssertEqual(pub, KNOWN_PUBKEY, "recovered pubkey");
            });

            Test("sha256_golden", () =>
            {
                using var ctx = new Ufsecp.Ufsecp();
                var digest = ctx.Sha256(Array.Empty<byte>());
                AssertEqual(digest, SHA256_EMPTY, "SHA-256 empty");
            });

            Test("addr_p2wpkh", () =>
            {
                using var ctx = new Ufsecp.Ufsecp();
                var addr = ctx.AddrP2wpkh(KNOWN_PUBKEY, 0);
                Assert(addr.StartsWith("bc1q"), $"Expected bc1q..., got {addr}");
            });

            Test("wif_roundtrip", () =>
            {
                using var ctx = new Ufsecp.Ufsecp();
                var wif = ctx.WifEncode(KNOWN_PRIVKEY, true, 0);
                var (key, comp, net) = ctx.WifDecode(wif);
                AssertEqual(key, KNOWN_PRIVKEY, "WIF privkey");
                Assert(comp == 1, "compressed");
                Assert(net == 0, "mainnet");
            });

            Test("ecdh_symmetric", () =>
            {
                using var ctx = new Ufsecp.Ufsecp();
                var k2 = HexToBytes("0000000000000000000000000000000000000000000000000000000000000002");
                var pub1 = ctx.PubkeyCreate(KNOWN_PRIVKEY);
                var pub2 = ctx.PubkeyCreate(k2);
                var s12 = ctx.Ecdh(KNOWN_PRIVKEY, pub2);
                var s21 = ctx.Ecdh(k2, pub1);
                AssertEqual(s12, s21, "ECDH symmetric");
            });

            Test("error_path", () =>
            {
                using var ctx = new Ufsecp.Ufsecp();
                bool threw = false;
                try { ctx.SeckeyVerify(new byte[32]); }
                catch { threw = true; }
                Assert(threw, "zero key should throw");
            });

            Test("ecdsa_deterministic", () =>
            {
                using var ctx = new Ufsecp.Ufsecp();
                var sig1 = ctx.EcdsaSign(MSG32, KNOWN_PRIVKEY);
                var sig2 = ctx.EcdsaSign(MSG32, KNOWN_PRIVKEY);
                AssertEqual(sig1, sig2, "RFC 6979 deterministic");
            });

            Console.WriteLine(new string('=', 60));
            Console.WriteLine($"  C# smoke test: {passed} passed, {failed} failed");
            Console.WriteLine(new string('=', 60));
            Environment.Exit(failed > 0 ? 1 : 0);
        }

        // ── Helpers ─────────────────────────────────────────────────────

        static void Assert(bool cond, string msg)
        {
            if (!cond) throw new Exception($"Assertion failed: {msg}");
        }

        static void AssertEqual(byte[] a, byte[] b, string label)
        {
            if (!a.SequenceEqual(b))
                throw new Exception($"{label}: {BitConverter.ToString(a)} != {BitConverter.ToString(b)}");
        }

        static byte[] HexToBytes(string hex)
        {
            var bytes = new byte[hex.Length / 2];
            for (int i = 0; i < bytes.Length; i++)
                bytes[i] = Convert.ToByte(hex.Substring(i * 2, 2), 16);
            return bytes;
        }
    }
}

File: bindings/java/tests/SmokeTest.java
// ============================================================================
// UltrafastSecp256k1 — Java Binding Smoke Test (Golden Vectors)
// ============================================================================
// Verifies FFI/JNI boundary correctness using deterministic known-answer tests.
// Runs in <2 seconds.
//
// Usage:
//   javac -d out SmokeTest.java && java -cp out SmokeTest
// ============================================================================

package com.ultrafast.ufsecp.tests;

import com.ultrafast.ufsecp.Ufsecp;

public class SmokeTest {

    static int passed = 0, failed = 0;

    // ── Golden Vectors ──────────────────────────────────────────────────

    static final byte[] KNOWN_PRIVKEY = hexToBytes(
        "0000000000000000000000000000000000000000000000000000000000000001");

    static final byte[] KNOWN_PUBKEY = hexToBytes(
        "0279BE667EF9DCBBAC55A06295CE870B07029BFCDB2DCE28D959F2815B16F81798");

    static final byte[] KNOWN_XONLY = hexToBytes(
        "79BE667EF9DCBBAC55A06295CE870B07029BFCDB2DCE28D959F2815B16F81798");

    static final byte[] SHA256_EMPTY = hexToBytes(
        "E3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B934CA495991B7852B855");

    static final byte[] MSG32 = new byte[32];
    static final byte[] AUX32 = new byte[32];

    // ── Helpers ─────────────────────────────────────────────────────────

    static void check(String name, Runnable fn) {
        try {
            fn.run();
            System.out.println("  [OK] " + name);
            passed++;
        } catch (Exception e) {
            System.out.println("  [FAIL] " + name + ": " + e.getMessage());
            failed++;
        }
    }

    static void assertTrue(boolean cond, String msg) {
        if (!cond) throw new RuntimeException("Assertion failed: " + msg);
    }

    static void assertArrayEquals(byte[] a, byte[] b, String label) {
        if (a.length != b.length)
            throw new RuntimeException(label + ": length " + a.length + " != " + b.length);
        for (int i = 0; i < a.length; i++) {
            if (a[i] != b[i])
                throw new RuntimeException(label + ": differ at index " + i);
        }
    }

    static byte[] hexToBytes(String hex) {
        int len = hex.length() / 2;
        byte[] out = new byte[len];
        for (int i = 0; i < len; i++) {
            out[i] = (byte) Integer.parseInt(hex.substring(i * 2, i * 2 + 2), 16);
        }
        return out;
    }

    // ── Tests ───────────────────────────────────────────────────────────

    public static void main(String[] args) {
        System.out.println("UltrafastSecp256k1 Java Smoke Test");
        System.out.println("============================================================");

        try (Ufsecp ctx = new Ufsecp()) {

            check("ctx_create_abi", () -> {
                int abi = ctx.abiVersion();
                assertTrue(abi >= 1, "ABI " + abi + " < 1");
            });

            check("pubkey_create_golden", () -> {
                byte[] pub = ctx.pubkeyCreate(KNOWN_PRIVKEY);
                assertTrue(pub != null, "pubkeyCreate returned null");
                assertArrayEquals(pub, KNOWN_PUBKEY, "compressed pubkey");
            });

            check("pubkey_xonly_golden", () -> {
                byte[] xonly = ctx.pubkeyXonly(KNOWN_PRIVKEY);
                assertTrue(xonly != null, "pubkeyXonly returned null");
                assertArrayEquals(xonly, KNOWN_XONLY, "x-only pubkey");
            });

            check("ecdsa_sign_verify", () -> {
                byte[] sig = ctx.ecdsaSign(MSG32, KNOWN_PRIVKEY);
                assertTrue(sig != null && sig.length == 64, "sig null or wrong length");
                boolean ok = ctx.ecdsaVerify(MSG32, sig, KNOWN_PUBKEY);
                assertTrue(ok, "valid sig rejected");

                // Mutated → fail
                byte[] bad = sig.clone();
                bad[0] ^= 0x01;
                boolean fail = ctx.ecdsaVerify(MSG32, bad, KNOWN_PUBKEY);
                assertTrue(!fail, "mutated sig accepted");
            });

            check("schnorr_sign_verify", () -> {
                byte[] sig = ctx.schnorrSign(MSG32, KNOWN_PRIVKEY, AUX32);
                assertTrue(sig != null && sig.length == 64, "sig null or wrong length");
                boolean ok = ctx.schnorrVerify(MSG32, sig, KNOWN_XONLY);
                assertTrue(ok, "valid schnorr sig rejected");
            });

            check("ecdsa_recover", () -> {
                var rec = ctx.ecdsaSignRecoverable(MSG32, KNOWN_PRIVKEY);
                assertTrue(rec != null, "recoverable sign returned null");
                byte[] pub = ctx.ecdsaRecover(MSG32, rec.signature, rec.recoveryId);
                assertTrue(pub != null, "recovery returned null");
                assertArrayEquals(pub, KNOWN_PUBKEY, "recovered pubkey");
            });

            check("sha256_golden", () -> {
                byte[] digest = Ufsecp.sha256(new byte[0]);
                assertTrue(digest != null, "sha256 returned null");
                assertArrayEquals(digest, SHA256_EMPTY, "SHA-256 empty");
            });

            check("addr_p2wpkh", () -> {
                String addr = ctx.addrP2wpkh(KNOWN_PUBKEY, 0);
                assertTrue(addr != null && addr.startsWith("bc1q"),
                    "Expected bc1q..., got " + addr);
            });

            check("wif_roundtrip", () -> {
                String wif = ctx.wifEncode(KNOWN_PRIVKEY, true, 0);
                assertTrue(wif != null, "wifEncode returned null");
                var decoded = ctx.wifDecode(wif);
                assertTrue(decoded != null, "wifDecode returned null");
                assertArrayEquals(decoded.privkey, KNOWN_PRIVKEY, "WIF privkey");
                assertTrue(decoded.compressed, "WIF compressed");
                assertTrue(decoded.network == 0, "WIF mainnet");
            });

            check("ecdh_symmetric", () -> {
                byte[] k2 = hexToBytes(
                    "0000000000000000000000000000000000000000000000000000000000000002");
                byte[] pub1 = ctx.pubkeyCreate(KNOWN_PRIVKEY);
                byte[] pub2 = ctx.pubkeyCreate(k2);
                byte[] s12 = ctx.ecdh(KNOWN_PRIVKEY, pub2);
                byte[] s21 = ctx.ecdh(k2, pub1);
                assertArrayEquals(s12, s21, "ECDH symmetric");
            });

            check("error_path", () -> {
                byte[] zeroes = new byte[32];
                boolean threw = false;
                try {
                    ctx.pubkeyCreate(zeroes);
                } catch (Exception e) {
                    threw = true;
                }
                assertTrue(threw, "zero key should throw/fail");
            });

            check("ecdsa_deterministic", () -> {
                byte[] sig1 = ctx.ecdsaSign(MSG32, KNOWN_PRIVKEY);
                byte[] sig2 = ctx.ecdsaSign(MSG32, KNOWN_PRIVKEY);
                assertArrayEquals(sig1, sig2, "RFC 6979 deterministic");
            });

        }

        System.out.println("============================================================");
        System.out.println("  Java smoke test: " + passed + " passed, " + failed + " failed");
        System.out.println("============================================================");
        System.exit(failed > 0 ? 1 : 0);
    }
}
